### ISSUE TITLE
Various OpenCL Optimizations

### ISSUE THREAD

**Issue by GordonHarley** [buddhi1980/mandelbulber2#447]
##### System information (version)
- mandelbulber2 => 2.12.1.0
- Operating System / Platform => Win 10 x64
- Compiler => Binary Release

##### Detailed description (Enhancement)
Use fma instead of x * y + z where possible.  

##### Reasoning
fma is as fast as or faster than the above version, and additionally retains higher precision according to OpenCL spec since the steps are performed without intermediate precision loss. I'm in the process of converting over to attempt to identify the speedup level, but it is recommended by the spec.   

##### Steps to reproduce
N/A.  Performance testing in progress locally.

**Comment by GordonHarley** [comment]
Conversions to  FMA and replacement of branches designed to clamp in one direction to 1.0f or 0.0f with fmin and fmax, loop unrolling of some inner loops, and clamp when a range of 1.0f >= x >= 0.0f  in two files brought render times to .1s with default settings and under 10s without reflections / refractions but multiple volume lights, fog, and glow to an average of 16s with most under 10s @ 1080p.     I'll work on this and place them in my repository clone so you can use them (or not) at your leisure when I'm satisfied they aren't causing any artifacts, but it looks good so far aside from refraction, which needs a higher trace depth to work properly and hasn't been modified much yet.  Refraction artifacts were similar to those I was getting before, so they may have been related to the fractal used to test.  

All rendered on high quality settings except where mentioned. 

16.4s
![4 placed lights, non penetrating, soft shadows, low DE](https://e9fbffffff.files.wordpress.com/2018/02/dodecahedron_ocl_test_4_lights_shadows_16-4s.png)

<10s
![Glow, placed light, penetrating global light.  Higher DE number than normal to check accuracy](https://e9fbffffff.files.wordpress.com/2018/02/dodecahedron_ocl_test_no_reflect_higher_de.png)

~8min
![Two volumetric lights, fog, glow, raytracing, low DE value (0.1 or so) and raised accuracy & smoothness](https://e9fbffffff.files.wordpress.com/2018/02/dodecahedron_ocl_test_no_trans_full_8bit_8min.png)

~23s
![40 random size 5 lights, more distant fog](https://e9fbffffff.files.wordpress.com/2018/02/dodecahedron_40_random_lights_23-9s.png) 

Example with reflection and refraction (refraction not showing very prominently) from a couple days ago with about the same quality settings prior to changing any OpenCL.  ~25 minutes at lower resolution.  
![Pre-optimize](https://e9fbffffff.files.wordpress.com/2018/02/fractdeca.jpg) 

I'm willing to continue optimizing where possible and implement additional items if I have time, just posting these as an example since I saw very high speed increases in items related to files I changed.

**Comment by buddhi1980** [comment]
Hi
I'm happy to see ongoing optimization. 
I always try to focus on designing algorithms and during that it's difficult to optimize them. Thinking about optimization reduces my creativity. I'm glad that you are trying to improve the code.
I will apply all optimization which you do if they will not change results of algorithms and will not reduce dramatically readability of OpenCL code (which is crucial because of lack of debugging tools).

**Comment by GordonHarley** [comment]
Hi, 
Yeah, I know the feeling.   I worked on a large compiler system for years and gained a knack for reading the often-unclear manuals (the opencl ones are worse than ISO C or C++ IMO) and spotting optimization opportunities that aren't obvious at first glance.  The main issue I've run across over the years is that compilers will fail to implement "safe" optimizations if the code isn't in the exact layout they expect.  They might succeed on changing x * y + z => fma(x, y, z) but fail on x * y - z => fma(x, y, -z) for example.  If you don't find that hard to read that's likely to be the worst things will get.  I've been purposefully avoiding functions which lose precision since we're using 32 bit float to begin with.  Once I finish testing I'll send a couple of files along so you can let me know if you think something is hard to read, but I doubt you'll have a problem with readability.  As for changing results, I think it would be easy enough for me to throw together a test suite which runs the current version and  the modified one and flags any precision changes over a certain point for review. 

The debug capabilities are pretty bad;   the last time I installed Intel's OpenCL toolchain it crashed visual studio on load 3 out of 4 times and I had to remove it, and that's the only decent (on paper) debugger I've seen for the language; CPU debugging still works with standard debugging tools, it's just difficult.  Luckily the code can be swapped out and the program checked for valid behavior very quickly compared to needing a full rebuild.   

I'm interested in algorithms too, doing safe optimizations just gives me a way to read / understand the codebase before I write anything large.   

In terms of precision, which fractal zooms always need more of than you can get, I was going to bring up doing a fixed point GPU library, or finding one if it exists already.  Intel's FPGA board which will probably cost a small fortune is faster at fixed point than it is at float according to their white paper;  the full speed might not carry over to GPU but by scaling from smaller to larger types as needed a balance can be achieved there.   That's something to worry about more when everything else is where you want  it though.  

Thanks, and great job on putting this project together.   For years I was too busy and wondering why there wasn't an OpenCL or GLSL fractal renderer as it seems like an obvious choice.  

Have a good evening, 
Gordon

**Comment by zebastian** [comment]
@GordonHarley Thanks gordon, we could really need some educated eyes on this topic. 
The opencl code is pretty raw since it just emerged and there probably exist a couple of performance enhancements.
IMHO mad(...) is more performant / less accurate than fma(...). But that would need to be tested.

If you find some inperformances in the formula opencl code, let me know, so i can adopt the source cpp code of the formulas and the transpiler "parseToOpenCL" in the file: https://github.com/buddhi1980/mandelbulber2/blob/master/mandelbulber2/tools/populateUiInformation.php

The best debugger for opencl i could find is this one: https://gpuopen.com/compute-product/codexl/
I did setup mandelbulber with this and was able to run kernels and debug the performance of the assembled code. Worked quite well and documentation and video tutorials were really helpful.

If you have any questions let us know and we will be glad to help. Cheers!

**Comment by GordonHarley** [comment]
Thanks.  The transpiler is cool for this purpose.  Normally I'd suggest doing this as an optimization pass but I haven't found a good way of getting opencl code back out of SPIR / LLVM (only gpu assembly), and working with the compiler IR at that level is far more complicated than I'd recommend dealing with unless needed.  One item I didn't see in there that was mentioned as an enhancement somewhere in one of the GPU backends was to convert x*x*x where x > 2 to native_powr where possible.  

I've been avoiding mad because the OpenCL spec says that it may have "infinite error" for any inputs, but particularly that its definition allows mad(a, b, -a*b) to return nearly any value.  I think the later specs (2.0 and up) clarify that they're truncating an intermediate or using half precision, but it looks like I was wrong on fma for AMD cards:

They say: 
> The MAD instruction is an IEEE-compliant multiply followed by an IEEE-compliant add; it has the same accuracy as two separate MUL/ADD operations. No special compiler flags are required for the compiler to convert separate MUL/ADD operations to use the MAD instruction.

...and fma is listed as 4 cycles with mad being 4 *per* cycle, since they interpreted the spec (what does infinite precision mean on a finite precision machine exactly) to mean fma should be implemented in double precision which they're slow at.  It's faster than a double mul / add pair, but slower than an f32 mul / add.  I'm going to switch uses over unless we find somebody hitting problems with nVidia.   Then you have the issue of it downclocking Intel CPUs and slowing down other instructions.    

Oh boy I love architecture level silliness...  😈 

Another one with some hard numbers (from an AMD manual, but OpenCL mentions control flow statements negatively as well so I would bet that similar applies to NVidia cards.  I'm ignoring CPU for now.  This would be more difficult to do in the transpiler (it really needs to be compiler level) but I've been doing similar in the engine files I'm working on.   The fact that they recommend these indicates the compiler won't necessarily optimize them (or is known not to).   

BTW this is one case where I know lots of people consider ternary statements less readable, but I think the nearly 60x speedup on a single change justifies it.  

> 
> Use predication rather than control-flow.  The predication allows the GPU to execute both paths of execution in parallel, which can be faster than attempting to minimize the work through clever control-flow. The reason for this is that if no memory operation exists in a ?: operator (also called a 
> ternary operator), this operation is translated into a single cmov_logical instruction, which is executed in a single cycle. An example of this is: 
> ```
> If (A>B) {
>  C += D;
> } else {
>  C -= D;
> }
> ```
> Replace this with:
> ```
> int factor = (A>B) ? 1:-1;
> C += factor*D;
> ```
> In the first block of code, this translates into an IF/ELSE/ENDIF sequence of CF clauses, each taking ~40 cycles. The math inside the control flow adds two cycles if the control flow is divergent, and one cycle if it is not. This code executes in ~120 cycles.
> 
> 

> **7.10.7.2  Remove Conditional Assignments**
> A conditional of the form “if-then-else” generates branching and thus generates one or more clauses. Use the select() function to replace these structures with conditional assignments that do not cause branching. For example:
> ```
> if(x==1) r=0.5;
> if(x==2) r=1.0;
> ```
> becomes
> ```
> r = select(r, 0.5, x==1);
> r = select(r, 1.0, x==2);
> ```
> Note that if the body of the if statement contains an I/O, the if statement cannot be eliminated.

And similarly: 

> **7.10.7.3  Bypass Short-Circuiting**
> A conditional expression with many terms can compile into a number of clauses due to the C-language requirement that expressions must short circuit. To prevent this, move the expression out of the control flow statement. For example:
> `if(a&&b&&c&&d){...}`
> becomes
> ```
> bool cond = a&&b&&c&&d;
> if(cond){...}
> ```
> The same applies to conditional expressions used in loop constructs (do, while, for).

Lots of these (especially the last one) are non-intuitive but the 118 cycle hit for using if / else instead of the short form is a great example.   There's also mention of nested if statements causing generation of 2^k conditional clauses where k is the nesting level, and if they're as badly performant as the first example mentions that's going to get ugly fast.   😺 

Just some more to chew on.  Now, back to doing something other than reading.

**Comment by buddhi1980** [comment]
As mad instructions are automatically used by compiler during optimization, we could leave this for the compiler. 
Other stuff like cond?result1:result2 or select() I really like. 
One big optimizations about branching which I already did was to use macros instead of if() statements. It additionally reduced compiled code size.

**Comment by GordonHarley** [comment]
Glad you guys don't mind me wandering in btw, think you'll like this:  
Got a speedup with several changes on the case I'm using as standard right now.  88s -> 50.3s;  did not expect that big a difference.   Would somebody like to run this fractal on OpenCL full engine?  I'd like to get a general idea of what it does on other machines.    I know the video card will make a difference, but I'll have some code in my clone of this as soon as I test with other fractal types.  There's enough modified right now that I don't want to change much more without a checkin / review.  
a) Getting some better idea of how fast this is on another card with unmodified engine. 
b) Throwing it on another git repo so somebody else can test the changes and make sure I haven't broken anything or slowed down performance on other cards.  The AMD stuff holds true back to the 5800 series according to their docs, and matches the OpenCL spec (except fma vs. mad), but if NVidia slows down there's a problem.
I'll fork the repository off so someone can code review as soon as I'm through double checking things with more formulas first.  

# Mandelbulber settings file
# version 2.12
# only modified parameters
[main_parameters]
ambient_occlusion_enabled true;
ambient_occlusion_mode 1;
ambient_occlusion_quality 10;
aux_light_colour_2 e900 d400 f500;
aux_light_enabled_1 true;
aux_light_enabled_2 true;
aux_light_intensity_1 2.567727732262943e-05;
aux_light_intensity_2 1.221494628743264e-05;
aux_light_intensity_3 1.365302555742003e-05;
aux_light_intensity_4 1.031936600863613e-05;
aux_light_position_1 0.5019758377874567 -0.4113407240945477 -0.06662180915801066;
aux_light_position_2 -0.2300461157773431 -0.5145412151991522 0.06629989930567635;
aux_light_position_3 -0.5788332542695491 -0.05286809340830623 -0.04458509417213854;
aux_light_position_4 -0.5118948952218076 -0.1755078765867487 0.2439198206037198;
aux_light_visibility_size 2;
aux_light_volumetric_intensity_1 0.0645654;
aux_light_volumetric_intensity_2 0.00849541;
brightness 1.7;
camera -0.3937175576110843 -0.2678441957937392 0.1244573756317752;
camera_distance_to_target 1.33856241478308;
camera_rotation -20.19608301307607 9.73630335531861 0;
camera_top -0.05838387481468789 -0.158716207239856 0.9855965141583134;
contrast 1.14;
DE_factor 0.1;
flight_last_to_render 0;
formula_1 10;
gamma 1.78;
glow_color_1 c100 8800 7100;
glow_color_2 ee00 dc00 f300;
glow_intensity 1;
hdr true;
image_height 1200;
image_width 1920;
keyframe_last_to_render 0;
mat1_fractal_coloring_algorithm 1;
mat1_fresnel_reflectance true;
mat1_is_defined true;
penetrating_lights false;
raytraced_reflections true;
shadows_cone_angle 15;
target 0.06174365085967942 0.9703243187628627 0.3508268864077602;
volumetric_fog_colour_1_distance 0.2;
volumetric_fog_colour_2_distance 0.4;
volumetric_fog_distance_factor 0.7;
[fractal_1]
IFS_abs_x true;
IFS_abs_y true;
IFS_abs_z true;
IFS_direction_0 2.62 1 -1.62;
IFS_direction_1 -1.62 2.62 1;
IFS_direction_2 1 -1.62 2.62;
IFS_enabled_0 true;
IFS_enabled_1 true;
IFS_enabled_2 true;
IFS_offset 1 1 1;
IFS_rotation_enabled true;
IFS_scale 2.62;
[fractal_2]
IFS_rotation_enabled true;
[fractal_3]
IFS_rotation_enabled true;
[fractal_4]
IFS_rotation_enabled true;

Edit:  Down to 48s on second & third run due to caching most likely.

**Comment by buddhi1980** [comment]
3m 63s on nvidia gtx 1050

**Comment by buddhi1980** [comment]
If you create fork or branch, I can test it after optimization.

**Comment by mclarekin** [comment]
"Glad you guys don't mind me wandering in btw"" you are more than welcome,
and we welcome what you are doing. Team members have limited available time
to work on the project and there is still a huge amount of things to be
evaluated and/or implemented (plus still a few bugs to solve).

Having someone, with the expertise, to look over the code for optimization
is much appreciated. Plus you are explaining things well, so even I can get
the drift of what you are doing :)

I will run your settings on my old AMD gpu, the next time i am in Windows.

**Comment by GordonHarley** [comment]
Thanks.  

I may backup and remove the changes that could have broken some non IFS for the first checkin.  Found it earlier, and odd behavior :  it renders 1-4 blocks of the fractal then freezes.  After I find that I'll checkin.  BTW I think -cl-single-precision-constant may be unsupported according to some AMD docs.

**Comment by buddhi1980** [comment]
I recommend to do step-by-step changes in kernels. It's really difficult to find errors. If you test every change separately or trace changes history, you will be able to find what is wrong. 
Only tool which is really reliable for testing is to put printf() instructions inside the kernel code. With this you could be able to find where is infinite loop.
You should create fork for your work, then we can help you in making changes and testing.

**Comment by GordonHarley** [comment]
Will do.  I've got the fork made, going to change one file at a time as I check in, sometime in the next day or two.  

I'll track it down pretty quickly once I start swapping files out and throw some debug prints into the culprit.   I always forget you can call printf from an opencl program so thanks for that.  

More information as events warrant.  👍  I'll let you know once some files are in.  Have some IRL things to do but shouldn't be too long.

**Comment by mclarekin** [comment]
AMD radeon R9 M265X 8min 52sec

**Comment by GordonHarley** [comment]
Hi guys, been a rough several weeks for me with some nasty health issues.  I'll be back to looking at this again soon, just didn't want you thinking I'd gone and forgotten  :-)

Thanks for numbers.  I'm on a Radeon Vega64 liquid (miraculously gotten before bitcoin spiked and when they had dipped to something approaching the original MSRP).   The other card I have is a 7970 but reinstalling it isn't possible because everything is now in a thermal / sound insulated case.  

Is there any standard benchmark for the program right now?   I settled on the one I posted for quick testing because it's easier for me to spot changes to the more geometric fractals and it takes long enough that it's easy to quantify speed changes, but it doesn't cover all features

**Comment by mclarekin** [comment]
@ Gordon, sorry about health issues, and hope you will be able to soon look
at the code again.

Is there any standard benchmark for the program right now?

well I think I have seen mention of benchmark,  but it is unlikely to cover
all the features.  I would guess that you may need to create some "feature
specific" benchmark settings, depending on what code is being optimzed.

**Comment by mancoast** [comment]
@GordonHarley hope all is well!  Did you get your changes into the upstream repo?

**Comment by buddhi1980** [comment]
As nothing happens here I'm closing this
